### PR TITLE
Added dot and cos merge modes

### PR DESCRIPTION
**Pull request by farizrahman4u**
Automatically merge-able version of #880

### CODE REVIEW

**Comment by farizrahman4u** [comment]
@fchollet  Using a general ```axis``` argument for both ```dot``` and ```concat``` merge modes,replacing the ```concat_axis``` will be a major change in the API and will break existing programs and examples (even if  they dont use ```concat``` mode) and will also fail all merge tests.

File: keras/layers/core.py [hunk 1]
```diff
@@ -276,20 +276,33 @@ def get_config(self):
...
+            if len(layers) > 2:
+                raise Exception(mode + " merge takes exactly 2 layers")
+            shape1 = layers[0].output_shape
+            shape2 = layers[1].output_shape
+            if mode == 'dot':
+                if type(dot_axes) == int:
+                    if dot_axes < 0:
+                        dot_axes = len(shape1) - 1
+                    dot_axes = [range(len(shape1) - dot_axes, len(shape2)), range(1,dot_axes + 1)]
+                for i in rage(len(dot_axes[0])):
```
Comment (fchollet) [inline keras/layers/core.py]: Typo: it's `range`. This means that this mode wasn't tested, and in fact, never run.

**Comment by fchollet** [comment]
Good job. Now it would be neat to have tests for these two modes: most likely a new method in `test_sequential_model.py`. It can be similar to `test_merge_concat` and `test_merge_sum`, but you can actually skip 80% of what these two do and go straight to just running a `Merge` with different modes: `dot` with `int` axes, `dot` with tuples axes, etc. Should be quick and short...

File: tests/auto/test_sequential_model.py [hunk 2]
```diff
@@ -185,7 +185,59 @@ def test_merge_concat(self):
...
+        left = Sequential()
+        left.add(Dense(input_dim=input_dim, output_dim=nb_hidden))
+        left.add(Activation('relu'))
+
+        right = Sequential()
+        left.add(Dense(input_dim=input_dim, output_dim=nb_hidden))
+        right.add(Activation('relu'))
+
+        model = Sequential()
+        model.add(Merge([left, right], mode='dot', dot_axes=1))
```
Comment (fchollet) [inline tests/auto/test_sequential_model.py]: Here you are only testing one parameter type of `dot_axes`. Since several types are accepted, they all need to be tested.

File: tests/auto/test_sequential_model.py [hunk 3]
```diff
@@ -185,7 +185,59 @@ def test_merge_concat(self):
...
+        model.fit([X_train, X_train], y_train, batch_size=batch_size, nb_epoch=nb_epoch, verbose=0, shuffle=False)
+
+        loss = model.evaluate([X_train, X_train], y_train, verbose=0)
+        print('loss:', loss)
+        if loss > 0.7:
+            raise Exception('Score too low, learning issue.')
+        preds = model.predict([X_test, X_test], verbose=0)
+        classes = model.predict_classes([X_test, X_test], verbose=0)
+        probas = model.predict_proba([X_test, X_test], verbose=0)
+        print(model.get_config(verbose=1))
```
Comment (fchollet) [inline tests/auto/test_sequential_model.py]: Everything below isn't necessary since these features do not depend on the merge mode and are being tested in a different test method.

**Comment by farizrahman4u** [comment]
Assertion error : tests/auto/test_tasks.py. I havent touched that file. Let me try restarting the Travis CI build.

**Comment by farizrahman4u** [comment]
```TestTasks.test_vector_clf```  is failing for some reason.

**Comment by fchollet** [comment]
It's possible that something changed in the Numpy install on Travis (or with the VM), which in turn changed the behavior of the Numpy RNG, which caused an accuracy test to fail. I will look into it.

**Comment by fchollet** [comment]
Let's try the following:
- `test_tasks.py`, line 35, replace the number of epochs by 15 (previously 12). That should most likely fix the failing test (if it doesn't, there might be a deeper issue).
- can you squash your series of 17 commits into 1-2 commits to make the commit history of the project clean and easy to follow?

**Comment by farizrahman4u** [comment]
@fchollet  

* Lets not include fixing ```test_tasks.py``` in this PR. Things will get messy.
* Squashing: If we are ready to merge, I could open a separate PR with a single commit. Are we ready to merge?

**Comment by fchollet** [comment]
> Are we ready to merge?

The style is still not conform, so you could fix the style. Otherwise I'll just clean it up myself.

I would recommend you install a PEP8 linter, such as https://pypi.python.org/pypi/pep8

Otherwise I believe we are ready to merge.

**Comment by farizrahman4u** [comment]
@fchollet 
> The style is still not conform, so you could fix the style.

Sorry, I am kind of 'white space blind' :-) 

>  Otherwise I'll just clean it up myself.

Thanks. But I will cleanup my mess

 > I would recommend you install a PEP8 linter, such as https://pypi.python.org/pypi/pep8

Thanks for the tip!

**Comment by farizrahman4u** [comment]
```
C:\Users\Fariz\Documents\GitHub\keras\dot\keras\keras\layers> pep8 --statistics -qq core.py
62      E501 line too long (101 > 79 characters)
```

@fchollet  Are we done?

**Comment by fchollet** [comment]
Yes, looks good to me. You can squash the commits and I will merge this.
